### DISCUSSION TITLE
Migrate from postgresql provider to mysql

### DISCUSSION THREAD

**Discussion by baptisteArno** [prisma/prisma#17622] [Q&A]
I'm thinking about migrating from a PostgreSQL provider to MySQL.

The first thing that I'm seeing is the fact that scalar types are not supported with MySQL. What would be the best way to handle this?

**Comment by nurul3101** [comment]
Hey @baptisteArno 👋 

Can you please elaborate on which scalar types you are referring to?
Do you mean using [scalar lists](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-scalar-lists-arrays#adding-items-to-a-scalar-list)?

**Reply by baptisteArno** [reply]
Exactly, yes!

**Reply by janpio** [reply]
I also replied on Twitter, but here is better:
https://twitter.com/janpio/status/1620099559849598976

> Meh, no lists in MySQL :/ 
>
> We might one day emulate those, but not today: https://github.com/prisma/prisma/issues/5227
>
> That puts an end to it without a refactoring. 
> 
> https://github.com/baptisteArno/typebot.io/blob/main/packages/db/prisma/schema.prisma does not actually have many (I see one String list, a few Json lists) - so would certainly be possible.

The suggestion for the `String` could be to either replace it with a relation (and a model behind it), a `Json` field that can contain an array of `String`s, or (worst) somehow concatenate into a `String`.

The `Json[]` could probably just become a `Json` and then the list is in the content itself - but probably that would also require some updates in the code.

**Comment by baptisteArno** [comment]
Ok, so I'm currently thinking about the migration script that will be executed when the database will change in production.

I'm thinking of something like:

```ts
import { PrismaClient as PostgresqlPrismaClient } from 'db/prisma-clients/postgresql'
import { PrismaClient as MysqlPrismaClient } from 'db/prisma-clients/mysql'

const postgresqlPrismaClient = new PostgresqlPrismaClient({
  datasources: { db: { url: process.env.POSTGRESQL_DATABASE_URL } },
})

const mysqlPrismaClient = new MysqlPrismaClient({
  datasources: { db: { url: process.env.MYSQL_DATABASE_URL } },
})

const main = async () => {
  const [
    accounts,
    answers,
    apiTokens,
    chatSessions,
    claimableCustomPlans,
    collaboratorsOnTypebots,
    coupons,
    credentials,
    customDomains,
    dashboardFolders,
    invitations,
    logs,
    memberInWorkspaces,
    publicTypebots,
    results,
    sessions,
    typebots,
    users,
    verificationTokens,
    webhooks,
    workspaces,
    workspaceInvitations,
  ] = await Promise.all([
    postgresqlPrismaClient.account.findMany(),
    postgresqlPrismaClient.answer.findMany(),
    postgresqlPrismaClient.apiToken.findMany(),
    postgresqlPrismaClient.chatSession.findMany(),
    postgresqlPrismaClient.claimableCustomPlan.findMany(),
    postgresqlPrismaClient.collaboratorsOnTypebots.findMany(),
    postgresqlPrismaClient.coupon.findMany(),
    postgresqlPrismaClient.credentials.findMany(),
    postgresqlPrismaClient.customDomain.findMany(),
    postgresqlPrismaClient.dashboardFolder.findMany(),
    postgresqlPrismaClient.invitation.findMany(),
    postgresqlPrismaClient.log.findMany(),
    postgresqlPrismaClient.memberInWorkspace.findMany(),
    postgresqlPrismaClient.publicTypebot.findMany(),
    postgresqlPrismaClient.result.findMany(),
    postgresqlPrismaClient.session.findMany(),
    postgresqlPrismaClient.typebot.findMany(),
    postgresqlPrismaClient.user.findMany(),
    postgresqlPrismaClient.verificationToken.findMany(),
    postgresqlPrismaClient.webhook.findMany(),
    postgresqlPrismaClient.workspace.findMany(),
    postgresqlPrismaClient.workspaceInvitation.findMany(),
  ])

  await mysqlPrismaClient.$transaction([
    mysqlPrismaClient.answer.createMany({
      data: answers,
    }),
    mysqlPrismaClient.apiToken.createMany({
      data: apiTokens,
    }),
    mysqlPrismaClient.user.createMany({
      data: users.map(user => ({
        ...user,
        onboardingCategories: user.onboardingCategories ?? undefined,
      })),
    }),
    // And so on...
  ])
}

main().then()
```

Is this a good way to go? I need it to be as fast as possible.

**Reply by janpio** [reply]
Relations are indeed the problem here - the order of the tables would need to be well defined and work for the foreign key checks (or you disable that first via a transaction).

I personally would separate the reading of the data and the re-importing, probably via JSON files. Then I would focus on getting the export working, and the import working separately - before then executing everything once. A quick check in between by looking at a `git diff` of the json files would give me certainty things are working properly before I run the real import.

(Of course there are also professional tools do to this at much larger scale - but I assume as you did not reach for those your amount of data is small enough to use a more manual approach.)

**Comment by baptisteArno** [comment]
I ended up creating a CLI that helps you migrate all your data to a new database powered by Prisma: https://github.com/baptisteArno/prisma-database-sync

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
